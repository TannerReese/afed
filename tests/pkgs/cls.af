# Test cls builtin package

# Erroneous statements
john.invalid_method = ``,
john % 23 = ``,
x * 5 = ``,

# Create constructors
Person: Class {
	clsname: "Person",
	THE_ANSWER: 42,
	CITY_CONSTANT: "New York",
	in_city p: (p.__data + " is in " + CITY_CONSTANT).print,

	__call p msg: (p.__data + " says '" + msg + "'").print,
    __str p: "Person(" + p.__data + ")",
	meet p1 p2: (p1.__data + " meets " + p2.__data).print,
},

NumStr: cls.Class {
    new n s: [n, s],
    nmb ns: ns 0, str ns: ns 1,
	nmb_pow ns p: ns.nmb ^ p,
    __call ns i: ns.__data i,

    __add ns1 ns2: ns1.new (ns1.nmb + ns2 0) (ns1 1 + ns2.str),
    __rmul ns i: ns.new (i * ns.nmb) (ns.str * i),
    __str ns: "NumStr(" + ns.nmb.str + ", '" + ns.str + "')",
},

# Test class constants
Person.clsname = ``,
Person.CITY_CONSTANT.print = ``,
Person.THE_ANSWER = ``,


# Test instances
john: Person "John",
mary: Person "Mary",
john "Goodbye" = ``,
Person.meet john mary = ``,
mary.meet john = ``,
mary.in_city = ``,

x: NumStr 2 "a", y: NumStr (-7) "c",
y.nmb_pow 2 = ``,
x + y = ``,
5 * x = ``,

