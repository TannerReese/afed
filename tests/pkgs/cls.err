line 4, column 23 of "cls.af" Eval Error: Cannot call method invalid_method on type class instance
line 5, column 13 of "cls.af" Eval Error: Binary operator % not implemented between Person(John) and 23
line 6, column 9 of "cls.af" Eval Error: Binary operator * not implemented between NumStr(2, 'a') and 5
3 Eval Errors encountered
