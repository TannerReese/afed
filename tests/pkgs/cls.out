# Test cls builtin package

# Erroneous statements
john.invalid_method = `Eval Error: Cannot call method invalid_method on type class instance`,
john % 23 = `Eval Error: Binary operator % not implemented between Person(John) and 23`,
x * 5 = `Eval Error: Binary operator * not implemented between NumStr(2, 'a') and 5`,

# Create constructors
Person: Class {
	clsname: "Person",
	THE_ANSWER: 42,
	CITY_CONSTANT: "New York",
	in_city p: (p.__data + " is in " + CITY_CONSTANT).print,

	__call p msg: (p.__data + " says '" + msg + "'").print,
    __str p: "Person(" + p.__data + ")",
	meet p1 p2: (p1.__data + " meets " + p2.__data).print,
},

NumStr: cls.Class {
    new n s: [n, s],
    nmb ns: ns 0, str ns: ns 1,
	nmb_pow ns p: ns.nmb ^ p,
    __call ns i: ns.__data i,

    __add ns1 ns2: ns1.new (ns1.nmb + ns2 0) (ns1 1 + ns2.str),
    __rmul ns i: ns.new (i * ns.nmb) (ns.str * i),
    __str ns: "NumStr(" + ns.nmb.str + ", '" + ns.str + "')",
},

# Test class constants
Person.clsname = `"Person"`,
Person.CITY_CONSTANT.print = `New York`,
Person.THE_ANSWER = `42`,


# Test instances
john: Person "John",
mary: Person "Mary",
john "Goodbye" = `John says 'Goodbye'`,
Person.meet john mary = `John meets Mary`,
mary.meet john = `Mary meets John`,
mary.in_city = `Mary is in New York`,

x: NumStr 2 "a", y: NumStr (-7) "c",
y.nmb_pow 2 = `49`,
x + y = `NumStr(-5, 'ac')`,
5 * x = `NumStr(10, 'aaaaa')`,

